### ISSUE TITLE
NullableArrays{T,1} vs Array{Nullable{T},1} and joins

### ISSUE THREAD

**Issue by amellnik** [JuliaStats/DataFrames.jl#1088]
With DataArrays I would do things like

```julia
using DataFrames
df = DataFrame(Fish = ["Jhun", "Amir", "Bob"], Mass = [1.5, 2.2, 1.1])  
df[:FormalName] = [Nullable(string(get(n), " sensei")) for n in df[:Fish]] #get would not be needed with DAs
df
```

to construct a second DataArray from another column.  Now when I do this I end up with a column `df[:FormalName]` which has type `Array{Nullable{String},1}` rather than the type of `df[:Fish]` which is `NullableArrays.NullableArray{String,1}`.  This isn't necessarily a problem, but arrays of Nullables break other methods from DataFrames including join:

```julia
df2 = DataFrame(
    Food = ["Pellets", "Pellets", "Flakes", "Flakes"],
    FormalName = ["Amir sensei", "Nadia sensei", "Jhun sensei", "Amir sensei"],
    Quanitity = [3,3,2,4]
)
join(df2, df, on=:FormalName, kind=:left)
```

```
Left_idx: [2,2,1]  #This is my own debugging note and does not normally print
Leftonly_idx: [3] #This is my own debugging note and does not normally print
LoadError: BoundsError: attempt to access 3-element NullableArrays.NullableArray{String,1} at index [[1,4,3]]
while loading In[3], in expression starting on line 6

 in throw_boundserror(::NullableArrays.NullableArray{String,1}, ::Tuple{Array{Int64,1}}) at .\abstractarray.jl:355
 in checkbounds at .\abstractarray.jl:284 [inlined]
 in _getindex at .\multidimensional.jl:270 [inlined]
 in getindex(::NullableArrays.NullableArray{String,1}, ::Array{Int64,1}) at .\abstractarray.jl:752
 in copy!(::Array{Any,1}, ::Base.Generator{Array{Any,1},DataFrames.##18#19{Array{Int64,1}}}) at .\abstractarray.jl:477
 in getindex(::DataFrames.DataFrame, ::Array{Int64,1}, ::Colon) at C:\Users\amellnik\.julia\v0.5\DataFrames\src\dataframe\dataframe.jl:296
 in #join#48(::Symbol, ::Symbol, ::Function, ::DataFrames.DataFrame, ::DataFrames.DataFrame) at C:\Users\amellnik\.julia\v0.5\DataFrames\src\abstractdataframe\join.jl:309
 in (::Base.#kw##join)(::Array{Any,1}, ::Base.#join, ::DataFrames.DataFrame, ::DataFrames.DataFrame) at .\<missing>:0
```

Is this just a bug with `join`, or should arrays of Nullables not be used in DataFrames?  In the latter case, is there any way to automatically promote these to NullableArrays?

Note: this example is fairly contrived -- I noticed it by taking substrings of Nullables using a getindex method exported by Query.jl as mentioned [here](https://github.com/davidanthoff/Query.jl/issues/61#issuecomment-251263098).

**Comment by nalimilan** [comment]
Working with `Array{Nullable}` should work, though it won't be optimal WRT performance. I'll have a look.

**Comment by nalimilan** [comment]
I have a fix at #1089. Unfortunately, that part of the code doesn't seem to be very well tested, so there may well be more subtle bugs like this. Please keep stress testing master and reporting any issues you may find!
